### ISSUE TITLE
wrong error message for deleted service units ("Access denied")

### ISSUE THREAD

**Issue by sourcejedi** [systemd/systemd#6450]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd-233-6.fc26.x86_64

### Used distribution

Fedora 26

### In case of bug report: Expected behaviour you didn't see

Error messages should be technically correct (even if they're too broad to actually identify a problem).

### In case of bug report: Unexpected behaviour you saw

"Access denied" message when the fault is a service unit that does not exist.

### In case of bug report: Steps to reproduce the problem

```
# systemctl edit --full systemd-suspend.service  # and exit the editor without making changes
# rm /etc/systemd/system/systemd-suspend.service # delete the override file
# systemctl start systemd-suspend.service
Failed to start systemd-suspend.service: Access denied
See system logs and 'systemctl status systemd-suspend.service' for details.
# systemctl status systemd-suspend.service
Failed to get properties: Access denied
```

This made me think systemd was denying access to all requests.  But actually it was failing with the wrong error only for requests which read `systemd-suspend.service`.

**Comment by sourcejedi** [comment]
Technically it's not a log message.  The journal is never consulted; `systemctl status` aborts early.  I can see an Access Denied error that would be generated in `mac_selinux_generic_access_check()`.  Although it sets a bus reply message "Failed to get file context on %s.", which I'm not seeing in output.

Because it's deliberately suppressed, sigh.

```
const char *bus_error_message(const sd_bus_error *e, int error) {

        if (e) {
                /* Sometimes, the D-Bus server is a little bit too verbose with
                 * its error messages, so let's override them here */
                if (sd_bus_error_has_name(e, SD_BUS_ERROR_ACCESS_DENIED))
                        return "Access denied";
```

I don't know how to resolve that.

**Comment by poettering** [comment]
Yeah, it's messy, the errors dbus spits out are just undigestale for users, they contain whole match strings and XML bits and we really don#t want to show all tha tin user facing UI...

What I don't get, how come mac_selinux_generic_access_check() returns an access error if the unit file doesn't exist? That appears to be the thing to fix here...

**Comment by keszybz** [comment]
I think this was fixed by 173471b771a52baa7f705019600956e875b0cb9a. When I follow your recipe to reproduce, `start` and `status` work without issues.

**Comment by sourcejedi** [comment]
I get the same result, even after `fixfiles restore && setenforce 1`.  I'll make sure to re-report if I notice it again :).

**Comment by jnpkrn** [comment]
Hello, maybe the issue would be worth reviving, since this is the
exact behaviour I've just observed with user service file, akin to:

```
$ systemctl --user status pulseaudio
● pulseaudio.service - Sound Service
     Loaded: loaded (/home/user/.config/systemd/user/pulseaudio.service; enabled; vendor preset: enabled)
[...]
$ rm ~/.config/systemd/user/pulseaudio.service
$ systemctl --user status pulseaudio
Failed to get properties: Access denied
```

I'd expect to be asked to reload the daemon as usual.
Current behaviour is indeed cryptic and unhelpful.

This was with:

```
$ systemctl --version
systemd 244 (v244.1-2.fc32)
```

Thanks for considering!

**Comment by sourcejedi** [comment]
Thanks. I re-tested my "steps to reproduce the problem" on `systemd-243.5-1.fc31.x86_64`, and I see the error happening now.

Based on strace, it looks like the same SELinux error.

`[pid 27159] <... recvmsg resumed>{msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\4\1s\0'\0\0\0org.freedesktop.DBus.Error.AccessDenied\0\7\1s\0\30\0\0\0org.freedesktop.systemd1\0\0\0\0\0\0\0\0\10\1g\0\1s\0\0J\0\0\0Failed to get file context on /etc/systemd/system/systemd-suspend.service.\0", iov_`

**Comment by sourcejedi** [comment]
Sorry, I missed the point.  If I try to reproduce using a `--user` unit instead as you mention (e.g. `pulseaudio.service`), I can reproduce your non-useful error messages on `systemd-243.6-1.fc31.x86_64`.

---

With a `--system` unit, I was getting an error again, but I was getting a *useful* error message.

```
# systemctl start systemd-suspend.service
Failed to start systemd-suspend.service: Access denied
See system logs and 'systemctl status systemd-suspend.service' for details.

# systemctl status systemd-suspend.service
Warning: The unit file, source configuration file or drop-ins of systemd-suspend.service changed on disk. Run 'systemctl daemon-reload' to reload units.
[...]
```

**Comment by cgzones** [comment]
Would something like this help to get a better error message?
```
diff --git a/src/core/selinux-access.c b/src/core/selinux-access.c
index abfab14dab..d59a8f8b34 100644
--- a/src/core/selinux-access.c
+++ b/src/core/selinux-access.c
@@ -234,7 +234,10 @@ int mac_selinux_generic_access_check(
                         if (!enforce)
                                 return 0;
 
-                        return sd_bus_error_setf(error, SD_BUS_ERROR_ACCESS_DENIED, "Failed to get file context on %s.", path);
+                        return sd_bus_error_setf(error,
+                                                 errno == ENOENT ? SD_BUS_ERROR_FILE_NOT_FOUND : SD_BUS_ERROR_ACCESS_DENIED,
+                                                 "Failed to get file context on %s.",
+                                                 path);
                 }
 
                 tclass = "service";
```

**Comment by jnpkrn** [comment]
@cgzones it's an improvement for sure.

Will the message about getting context still get displayed?
Note sure if mere existence is not more imporant than highlighting this
was found out when checking the SELinux context.

Is this codepath also used when compiled and/or used without SELinux support?  Any possibility of unification of externally observed behaviour in these cases
when the file is simply not there?

Don't take me bad, thanks for the effort, commenting is cheaper in comparison :-)

**Comment by cgzones** [comment]
> Will the message about getting context still get displayed?

Yes, it'll only change from
    `Failed to start systemd-suspend.service: Access denied`
to
    `Failed to start systemd-suspend.service: No such file or directory`
.

> Is this codepath also used when compiled and/or used without SELinux support? Any possibility of unification of externally observed behaviour in these cases
when the file is simply not there?

This codepath is only taken if systemd was compiled with SELinux support and SELinux is not disabled on the system.
In case SELinux is in permissive mode, (most) selinux functions return always 0 (success) instead of an error code.

Another option can be to reorder checks in `src/systemctl/systemctl.c`, maybe around 
https://github.com/systemd/systemd/blob/51115fdc42d3c7c142b107c24f46523d34a6b02f/src/systemctl/systemctl.c#L2912-L2923 ?

**Comment by cgzones** [comment]
@jnpkrn @sourcejedi 
I am not able to reproduce on Debian with systemd `245.6-2`.

With a service
```
[Unit]
Description=Test service
[Service]
Type=oneshot
ExecStart=/usr/bin/true
```

and a test script
```
#!/bin/sh
set -eux

UNIT_DIR=~/.config/systemd/user/
USER_INSTANCE=--user

mkdir -p ${UNIT_DIR}
cp test.service ${UNIT_DIR}
chmod 644 ${UNIT_DIR}/test.service
systemctl ${USER_INSTANCE} daemon-reload
systemctl ${USER_INSTANCE} status test || true
systemctl ${USER_INSTANCE} start test
rm ${UNIT_DIR}/test.service
systemctl ${USER_INSTANCE} status test || true
systemctl ${USER_INSTANCE} start test
```

I am getting
```
./test.sh 
+ UNIT_DIR=/root/.config/systemd/user/
+ USER_INSTANCE=--user
+ mkdir -p /root/.config/systemd/user/
+ cp test.service /root/.config/systemd/user/
+ chmod 644 /root/.config/systemd/user//test.service
+ systemctl --user daemon-reload
+ systemctl --user status test
_ test.service - Test service
     Loaded: loaded (/root/.config/systemd/user/test.service; static; vendor preset: enabled)
     Active: inactive (dead)

Jul 16 16:56:31 debianBullseye systemd[586]: Starting Test service...
Jul 16 16:56:31 debianBullseye systemd[586]: test.service: Succeeded.
Jul 16 16:56:31 debianBullseye systemd[586]: Finished Test service.
Jul 16 17:00:08 debianBullseye systemd[586]: Starting Test service...
Jul 16 17:00:08 debianBullseye systemd[586]: test.service: Succeeded.
Jul 16 17:00:08 debianBullseye systemd[586]: Finished Test service.
+ true
+ systemctl --user start test
+ rm /root/.config/systemd/user//test.service
+ systemctl --user status test
Unit test.service could not be found.
+ true
+ systemctl --user start test
Failed to start test.service: Unit test.service not found.
```

So a reasonable message `Unit test.service could not be found.`

**Comment by poettering** [comment]
Fixed by #23986 I guess.
